### PR TITLE
[dart-dio] Allow dynamic headers and add additional dio parameters

### PR DESCRIPTION
**Pull request by kuhnroyal**
* fixes some compile errors due to header params not being of type `String`
* add optional `extra` and `validateStatus` parameters from dio to API calls

Contains the same changes as https://github.com/OpenAPITools/openapi-generator/pull/6915 and fixes compile issues for dart-dio mentioned in https://github.com/OpenAPITools/openapi-generator/issues/8179.

Regarding the concerns in #6915 about missing changes to the dart generator: 
- dart generator does not expose the header map
- the `http.Client` only handles `String` headers where as `dio` can handle `dynamic` headers and does its own serialization of headers

### PR checklist
 
- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Pull Request title clearly describes the work in the pull request and Pull Request description provides details about how to validate the work. Missing information here may result in delayed response from the community.
- [x] If contributing template-only or documentation-only changes which will change sample output, [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) beforehand.
- [x] Run the shell script `./bin/generate-samples.sh`to update all Petstore samples related to your fix. This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit as it would merge with master. These must match the expectations made by your contribution. You may regenerate an individual generator by passing the relevant config(s) as an argument to the script, for example `./bin/generate-samples.sh bin/configs/java*`. For Windows users, please run the script in [Git BASH](https://gitforwindows.org/).
- [x] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master`
- [x] Copy the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) to review the pull request if your PR is targeting a particular programming language.

CC @swipesight (2018/09) @jaumard (2018/09) @josh-burton (2019/12) @amondnet (2019/12) @sbu-WBT (2020/12) @kuhnroyal (2020/12) @agilob (2020/12)

### CODE REVIEW

File: modules/openapi-generator/src/main/resources/dart-dio/api.mustache [hunk 1]
```diff
@@ -23,14 +23,18 @@ class {{classname}} {
...
+        ValidateStatus validateStatus,
         ProgressCallback onSendProgress,
         ProgressCallback onReceiveProgress,
     }) async {
         final String _path = '{{{path}}}'{{#pathParams}}.replaceAll('{' r'{{baseName}}' '}', {{{paramName}}}.toString()){{/pathParams}};
 
         final Map<String, dynamic> queryParams = {};
-        final Map<String, String> headerParams = Map.from(headers ?? {});
+        final Map<String, dynamic> headerParams = {
+            if (headers != null) ...headers,
```
Comment (agilob) [inline modules/openapi-generator/src/main/resources/dart-dio/api.mustache:36]: never seen this syntax before. basically `headerParams` gets content of `headers` if `headers` is initialised?

**Review by agilob — APPROVED** [review]

File: modules/openapi-generator/src/main/resources/dart-dio/api.mustache (hunk 1 above)
Comment (kuhnroyal) [reply modules/openapi-generator/src/main/resources/dart-dio/api.mustache:36]: Ya basically like the ecma spread operator. Works for all collections.

**Review by josh-burton — APPROVED** [review]

**Comment by wing328** [comment]
@kuhnroyal I just rebased and merged https://github.com/OpenAPITools/openapi-generator/pull/6915. Can you please rebase this one to resolve the merge conflicts when you've time?

**Comment by kuhnroyal** [comment]
@wing328 On it.

**Comment by kuhnroyal** [comment]
@wing328 rebased

**Comment by wing328** [comment]
CircleCI failure not related to this change.

**Pull request merged**